### PR TITLE
[WIP] diagnostic rest api

### PR DESCRIPTION
**Pull request by ikitiki**
Diagnostic REST API provides profiling data via pprof, cluster and worker logs, cluster and controller statuses and configuration.

### CODE REVIEW

File: pkg/controller/postgresql.go [hunk 1]
```diff
@@ -146,86 +145,112 @@ func (c *Controller) processEvent(obj interface{}) error {
...
-		c.clusters[clusterName] = cl
-		c.stopChs[clusterName] = stopCh
-		c.clustersMu.Unlock()
+		func() {
+			defer c.clustersMu.Unlock()
+			c.clustersMu.Lock()
+
+			c.teamClusters[teamName] = append(c.teamClusters[teamName], clusterName)
+			c.clusters[clusterName] = cl
+			c.stopChs[clusterName] = stopCh
```
Comment (alexeyklyukin) [inline pkg/controller/postgresql.go]: why a stop channel per cluster?

File: pkg/controller/postgresql.go [hunk 2]
```diff
@@ -146,86 +145,112 @@ func (c *Controller) processEvent(obj interface{}) error {
...
-		delete(c.stopChs, clusterName)
-		c.clustersMu.Unlock()
+		func() {
+			defer c.clustersMu.Unlock()
+			c.clustersMu.Lock()
+
+			delete(c.clusters, clusterName)
+			delete(c.stopChs, clusterName)
+			delete(c.clusterLogs, clusterName)
+			for i, val := range c.teamClusters[teamName] { // on relativel
```
Comment (alexeyklyukin) [inline pkg/controller/postgresql.go]: perhaps teamClusters should be a List?

File: pkg/controller/postgresql.go [hunk 3]
```diff
@@ -146,86 +145,112 @@ func (c *Controller) processEvent(obj interface{}) error {
...
+		lg.Infoln("Cluster has been deleted")
 	case spec.EventSync:
-		logger.Infof("Syncing of the %q cluster started", clusterName)
+		lg.Infoln("Syncing of the cluster started")
 
 		// no race condition because a cluster is always processed by single worker
 		if !clusterFound {
 			stopCh := make(chan struct{})
-			cl = cluster.New(c.makeClusterConfig(), c.KubeClient, *event.NewSpec, logger)
+			cl = cluster.New(c.makeClusterConfig(), c.KubeClient, *event.NewSpec, lg)
```
Comment (alexeyklyukin) [inline pkg/controller/postgresql.go]: that parts duplicates the code from Create, perhaps refactor it into a function?

File: pkg/util/ringlog/ringlog.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+type RingLogger interface {
+	Insert(level logrus.Level, t time.Time, workerID *uint32, clusterName spec.NamespacedName, message string) error
+	Walk() []*LogEntry
+}
+
+// RingLog is a capped logger with fixed size
+type RingLog struct {
+	size int
+	list *list.List
+	mu   *sync.RWMutex
```
Comment (alexeyklyukin) [inline pkg/util/ringlog/ringlog.go]: Does it make sense to embed *sync.RWMutex to be able to write r.Lock instead of r.mu.Lock?

Comment (ikitiki) [inline pkg/util/ringlog/ringlog.go]: good idea.

**Comment by alexeyklyukin** [comment]
👍

**Comment by alexeyklyukin** [comment]
Note: let's split such PRs into multiple smaller parts in the future, as in the current state it is pretty much unreviewable due to its size and different changes mixed up together.

**Pull request closed without merging**